Relocatable OCaml - test harness

dra27 opened the pull request:
This PR implements a test harness for Relocatable OCaml (`ocaml/RFCs#53`). The RFC defines "Relocatable" as meaning:
1. The compiler binaries are identical regardless of the installation _prefix_ (the `--prefix` argument passed to `configure`) or the working directory in which the compiler was built (the _build path_).
2. The resulting compiler distribution can be used from any disk location on the system without any further alteration to the binaries.
3. The resulting compiler distribution can be used from any disk location on the system without any further alteration to the user's shell environment.

Rather than building two compilers, the first property can be checked to a sufficient degree for Relocatable OCaml by searching the installed artefacts for _prefix_ and _build path_. The second and third properties can be simultaneously checked by moving the compiler from _prefix_ to a new location, and then running a series of tests to ensure that that compiler works.

Obviously, at present OCaml cannot be configured in a way which satisfies these properties, but it is of course possible to move a compiler to a new location and change the way the compiler is invoked in order to make it work. The harness in this PR implements this approach. Everything is thus tested with a combination of known failures and shims (such as setting `OCAMLLIB`, or explicitly invoking bytecode binaries with `ocamlrun`). The three further PRs which make up the proposed changes for Relocatable OCaml are based on top of this PR, and therefore update this test harness with the revised behaviour.

There are two key differences between this test harness and the main ocamltest-based testsuite:
1. it is central to this test that it runs on the compiler having been installed to the prefix it was configured with; and
2. the test focuses more on the operation and execution of the programs than of the actual programs themselves, requiring more control over the exact commands executed _to build the tests_ than ocamltest offers

A consequence of the "in-prefix" part is that this is _not_ a test that should be run by default and the fact it needs to operate outside the build tree has led to an additional harness, rather than additional features in ocamltest.

The harness itself has revealed various bugs not related to Relocatable OCaml at all (cf. `ocaml/flexdll#146`, `#13496`, `#13520`, `#13638`, `#13692` and `#13693`, in addition to a fault in the partial linker alluded to `#13692` which will be fixed in a subsequent PR)

The tests performed are covered in `testsuite/in_prefix/README.md`.

In terms of review, the PR includes the first commit of `#13728`, which means all the changes in `runtime/` can be ignored (as they all come from that commit). The next five commits alter the compiler:
- The compilation tests are _vastly_ simpler if they can use `Ccomp.call_linker` (I mean **really, really, simpler** - I tried!). However, in order to that, the test harness needs to be able to control slightly more precisely the value of `Config.standard_library` as interpreted by `Ccomp.call_linker`. Having tried various approaches, the least invasive to compiler-libs seems to be to generalise `Compmisc.init_path` via a new `Compmisc.reinit_path`. There are other ways of doing this, but this one I think is the simplest that doesn't involve duplicating code from `utils/ccomp.ml` directly in the test harness.
- `Test_ld_conf` tests the handling by both `ocamlrun` and `ocamlc` of `ld.conf`. This test needs to be able the content of the `search_path` variable in `Dll` for which I have added a `Dll.search_path` function.
- Both the test harness and one of the subsequent Relocatable PRs need to iterate over the code points of a UTF-8 `string`, and for this reason I've added `Misc.Stdlib.String.to_utf_8_seq` and, as always, I'm open to suggestions for a better name.
- The test harness needs to be able to parse `runtime-launch-info`, so I've exposed `Bytelink.read_runtime_launch_info` and its associated types.
- There are various parts of the tests for reproducibility which are clang-specific. The easiest way to do this seems to be to expose the `ocaml_cc_vendor` variable computed in `configure` as `Config.cc_vendor`

Those opening commits clearly change the compiler, and need to be reviewed as such. The last two alter the testsuite only, and I'd beseech be reviewed from a clarity perspective only! In particular, I have 12 branches based on top of this all of which alter this test harness, so please bear in mind that while adjustments are of course possible, they come at a cost. Note also that there are a few places where behaviour is not reachable or spelled out slightly obtusely: for example, the `Default_ocamlc` constructor includes an ignored `launch_method` value, and several functions have unused parameters escaped with `_`. These are all altered in subsequent PRs where the reason for having this differentiation becomes (hopefully) clearer.

As far as possible, the harness learns of the configuration statically, rather than by probing the installation. In particular, this allows errors in the installation process to be detected (i.e. if the toplevel test simply tested all the .cma files it found in libdir, it would not catch the failure to have installed one of them - hence `$(OTHERLIBRARIES)` is passed on the command line instead. Most things are statically placed in the `Config` module - `BINDIR` and `LIBDIR` are passed on the command line, because `Config.bindir` and `Config.standard_library` won't have the same meaning in subsequent PRs. Especially given that the need to be able to use `Ccomp.call_linker`, I've not written this test to be used in a cross-compiled manner by which I mean it can test an installation which is a cross compiler, but the test harness must itself be compiled with the same (host) configuration as the compiler it's testing.

The final commit plumbs the tests into CI - it (should!) be passing both on GitHub Actions here and has also passed precheck (although I can't at present schedule a run of this precise version of the PR, as precheck seems to be down). It's also passing an even wider text matrix including Cygwin and multiple different shebang/executable/static/minimal tests in dra27/ocaml#175.

The harness code itself is structured with utility functions in module `Harness` and information derived about the C compiler and platform in `Toolchain`. Module `Environment` (and in particular `Environment.run_process`) encapsulates the configuration of the configuration of the compiler installation. `Cmdline` and `Test_in_prefix` drive the entire process with `TestRelocation`, `TestBytecodeBinaries`, `TestDynlink`, `TestLinkModes`,`TestTopLevel` and `Test_ld_conf` containing the implementations of the six tests.

dra27 committed:
Add a test for the compiler "in-prefix"

dra27 committed:
Generalise Compmisc.init_path

Previously, Compmisc.init_path initialised the load path using
Config.standard_library, but this can now be altered via an optional
?standard_library argument. This is used internally when testing
compiler installations in order to allow Ccomp.call_linker to be used.

dra27 committed:
Plumb the in-prefix tests into CI

gasche committed:
[minor] printcmm: print atomic loads

gasche committed:
[testsuite] test the compilation of atomic functions

clef-men committed:
[refactor] Define [caml_atomic_cas] in terms of [caml_atomic_cas_field].

gasche committed:
[breaking] Change the prototype of [caml_atomic_cas_field].

This is a breaking change because this function was (unfortunately)
exposed outside CAML_INTERNALS, and is used by exactly one external
user, you guessed it:
  https://github.com/ocaml-multicore/multicore-magic/blob/360c2e829c9addeca9ccaee1c71f4ad36bb14a79/src/Multicore_magic.mli#L181-L185
  https://github.com/ocaml-multicore/multicore-magic/blob/360c2e829c9addeca9ccaee1c71f4ad36bb14a79/src/unboxed5/multicore_magic_atomic_array.ml#L36-L43

We chose to change the prototype to remain consistent with the naming
convention for the new caml_atomic_*_field primitives, which will be
added to support atomic record fields.

User code can easily adapt to this new prototype we are using, but not
in a way that is compatible with both old and new versions of
OCaml (not without some preprocessing at least).

Another option would be to expose

    int caml_atomic_cas_field(value obj, intnat fld, value, value)
    value caml_atomic_cas_field_boxed(value obj, value vfld, value, value)

but no other group of primitives in the runtime uses this _boxed
terminology, they instead use

    int caml_atomic_cas_field_unboxed(value obj, intnat fld, value, value)
    value caml_atomic_cas_field(value obj, value vfld, value, value)

and this would again break compatiblity -- it is not easier to convert
code to that two-version proposal, and not noticeably more efficient.

So in this case we decided to break compatibility (of an obscure,
experimental, undocumented but exposed feature) in favor of
consistency and simplificity of the result.

clef-men committed:
runtime: Add atomic primitives [caml_atomic_load_field], [caml_atomic_exchange_field] and [caml_atomic_fetch_add_field].

gasche committed:
lambda: Add support for new atomic primitives.

Uses of existing atomic primitives %atomic_foo, which act on
single-field references, are now translated into %atomic_foo_field,
which act on a pointer and an offset -- passed as separate arguments.

In particular, note that the arity of the internal Lambda primitive
    Patomic_load
increases by one with this patchset. (Initially we renamed it into
    Patomic_load_field
but this creates a lot of churn for no clear benefits.)

We also support primitives of the form %atomic_foo_loc, which
expects a pair of a pointer and an offset (as a single argument),
as we proposed in the RFC on atomic fields
  https://github.com/ocaml/RFCs/pull/39
(but there is no language-level support for atomic record fields yet)

Co-authored-by: Clément Allain <clef-men@orange.fr>

gasche committed:
Add tests for atomic record fields.

gasche committed:
[minor] Make test_untypeast.ml more readable.

clef-men committed:
Add module [Stdlib.Atomic.Loc].

gasche committed:
Add tests for [Atomic.Loc].

clef-men committed:
Reimplement [Atomic.t] using [Atomic.Loc.t].

clef-men committed:
atomic fields in ocamldoc

gasche committed:
[refactor] more general type for Env.mark_*_used

We want to use [mark_label_used] in a context where we cannot easily
find the label declaration, only the label description (from the
environment).

clef-men committed:
Forbid atomic fields in patterns.

gasche committed:
Changes

dra27 committed:
Add bytecode-only Cygwin check

dra27 committed:
Add optional Cygwin check to CI

dra27 committed:
Add a 'CI: Skip testsuite' label

Applying this label _before_ the testsuite has started in either GitHub
Actions or AppVeyor will cause the execution of the testsuite to be
skipped.

dra27 committed:
Add caml_sys_proc_self_exe primitive

caml_executable_name is always called in native startup and for all the
non-default bytecode linking mechanisms. Bytecode startup now always
calls caml_executable_name, and this value is stored along with
exe_name.

caml_sys_proc_self_exe returns this stored value as a string option. It
returns None if caml_executable_name is not implemented on a given
platform.

MisterDA commented on Makefile:
```diff
@@ -1986,6 +1986,19 @@ $(asmgen_OBJECT): $(asmgen_SOURCE)
+  $(addprefix testsuite/tools/,test_in_prefix.mli test_in_prefix.ml)
+test_in_prefix_LIBRARIES = otherlibs/unix/unix compilerlibs/ocamlcommon
+
+testsuite/tools/test_in_prefi%: CAMLC = $(BEST_OCAMLC) $(STDLIBFLAGS)
+
+testsuite/tools/test_in_prefix$(EXE): OC_BYTECODE_LINKFLAGS += -custom
+
+testsuite/tools/test_in_prefi%: CAMLOPT = $(BEST_OCAMLOPT) $(STDLIBFLAGS)
```
```suggestion
testsuite/tools/test_in_prefix%: CAMLC = $(BEST_OCAMLC) $(STDLIBFLAGS)

testsuite/tools/test_in_prefix$(EXE): OC_BYTECODE_LINKFLAGS += -custom

testsuite/tools/test_in_prefix%: CAMLOPT = $(BEST_OCAMLOPT) $(STDLIBFLAGS)
```

Why drop the leading `O` in `OCAMLC` and `OCAMLOPT`? or is it always spelled this way?

dra27 replied:
> Why drop the leading `O` in `OCAMLC` and `OCAMLOPT`? or is it always spelled this way?

@shindere definitely wants to change this (the names are just legacy from Caml Light and Caml Special Light!)... I _think_ the idea was to wait until the `Makefile`s are all merged... renaming Makefile variables is of course a tricky process to get right (all the recursive calls, etc.)

dra27 replied:
Makefile patterns cannot match the empty string, which is why the `x` is omitted - this rule is intended to match _both_ `testsuite/tools/test_in_prefix` and `testsuite/tools/test_in_prefix.opt` and the pattern `testsuite/tools/test_in_prefix%` doesn't match the first one.

cf. the same trick in `stdlib/Makefile`:
https://github.com/dra27/ocaml/blob/f28bfbcb232b017460cb0ba7f9ec8cf1cf773811/stdlib/Makefile#L93-L94
where the pattern is `%-launch-info` (to match both `runtime-launch-info` and `target_runtime-launch-info`)

MisterDA replied:
Interesting Makefile craziness... thanks!

MisterDA commented on runtime/caml/startup_aux.h:
```diff
@@ -18,6 +18,8 @@
 
 #ifdef CAML_INTERNALS
 
+#include <stdbool.h>
```
Is including `stdbool.h` fine here w.r.t (the absence of) namespaces and C++ ABI compatibility because it's guarded by `CAML_INTERNALS`?

dra27 replied:
Instinctively I'd say yes, but is there an immediately cleaner way I could put it? The other option would be just to use an `int`, of course 😉

MisterDA replied:
I think it's fine, we're not guarding things under `CAML_INTERNALS` with `extern "C"` for C++ compat either.

MisterDA commented on testsuite/in_prefix/README.md:
```diff
@@ -0,0 +1,157 @@
+This directory performs tests on the compiler after it has been installed to the
+prefix given to `configure`. The test requires at least the Unix library to have
+been compiled, and only works if bindir and libdir share a common root directory
+(e.g. /usr/bin and /usr/lib/ocaml share a common root of /usr, but /bin and
+/lib/ocaml do not share a common root directory). The test needs to be able to
+rename this common root directory by adding a suffix ".new". For this reasoa,
```
```suggestion
(e.g. `/usr/bin` and `/usr/lib/ocaml` share a common root of `/usr`, but `/bin` and
`/lib/ocaml` do not share a common root directory). The test needs to be able to
rename this common root directory by adding a suffix ".new". For this reason,
```

dra27 replied:
I had a thought (just to explain it), that putting these paths as `code` wasn't aiding readability (either rendered or as text) - i.e. I was finding "/usr/bin and /usr/lib/ocaml" easier to read (because each path is "introduced" by a slash anyway) than the version where the backticks are kinda getting in the way?

MisterDA replied:
fine by me :)

MisterDA commented on testsuite/in_prefix/README.md:
```diff
@@ -0,0 +1,157 @@
+   isn't available and native mode tests are skipped for
+   `--disable-native-compiler`):
+   - Bytecode/native standard compilation
+   - Bytecode compilation with -custom running on the static/shared runtime
+   - Bytecode compilation with -output-complete-exe on the static/shared runtime
+   - Bytecode/native compilation with -output-obj linked with the static/shared
+     runtime
+   - Bytecode/native compilation with -output-complete-obj linked with the
```
```suggestion
   - Bytecode compilation with `-custom` running on the static/shared runtime
   - Bytecode compilation with `-output-complete-exe` on the static/shared runtime
   - Bytecode/native compilation with `-output-obj` linked with the static/shared
     runtime
   - Bytecode/native compilation with `-output-complete-obj` linked with the
```

MisterDA commented on testsuite/in_prefix/README.md:
```diff
@@ -0,0 +1,157 @@
+
+### Executing installed bytecode binaries with `-vnum`
+
+This test looks for filenames match `flexlink*` and `ocaml*` in bindir. For each
```
```suggestion
This test looks for filenames match `flexlink*` and `ocaml*` in _bindir_. For each
```

MisterDA commented on testsuite/in_prefix/README.md:
```diff
@@ -0,0 +1,157 @@
+- a command resolvable in PATH
+- a relative invocation of the executable (i.e. `./prog`)
+- an implicit invocation where `.` is not in PATH (i.e. `prog`)
+- an implicit invocation but with `.` in PATH (i.e. `PATH=".:$PATH" prog)
```
```suggestion
- an implicit invocation but with `.` in PATH (i.e. `PATH=".:$PATH"` prog)
```

shindere commented:
David Allsopp (2024/12/27 02:02 -0800):
> @dra27 commented on this pull request.
> 
> 
> 
> > +testsuite/tools/test_in_prefi%: CAMLC = $(BEST_OCAMLC) $(STDLIBFLAGS)
> +
> +testsuite/tools/test_in_prefix$(EXE): OC_BYTECODE_LINKFLAGS += -custom
> +
> +testsuite/tools/test_in_prefi%: CAMLOPT = $(BEST_OCAMLOPT) $(STDLIBFLAGS)
> 
> > Why drop the leading `O` in `OCAMLC` and `OCAMLOPT`? or is it always spelled this way?
> 
> @shindere definitely wants to change this (the names are just legacy
> from Caml Light and Caml Special Light!)... I _think_ the idea was to
> wait until the `Makefile`s are all merged... renaming Makefile
> variables is of course a tricky process to get right (all the
> recursive calls, etc.)

I can only ocnfirm, @dra27 explained things in an excellent way.

dra27 commented:
Markdown fixed; and...

dra27 commented:
... rebased

dra27 committed:
Add Sys.runtime_executable

In native mode, same as Sys.executable_name, in bytecode, the path to
the interpreter executing Sys.executable_name, which may not be the same
from the same file.

dra27 committed:
Add Dll.search_path

Provides a copy of the DLL search path for the test harness.

dra27 committed:
Add Misc.Stdlib.String.to_utf_8_seq

dra27 committed:
Fix tools/check-symbol-names for i686 mingw-w64

OlivierNicole committed:
Add Dynarray.unsafe_to_iarray

Fixes #13793.

dra27 committed:
Expose Bytelink.read_runtime_launch_info

clementblaudeau committed:
[refactor merge] Create a Merge module + white space changes

clementblaudeau committed:
[refactor merge] Extracted helper functions

- [split_row_ids] and [unsafe_signature_subst]

clementblaudeau committed:
[refactor merge] Changed the return logic

- Typedtrees are no longer built inside [merge_constraint] but inside
[transl_with], which removes the need for a special approximation case: merging
always returns a module type, not a Typedtree. Changed [transl_with] to build
the Typedtree there.

- Removed the [real_ids] mechanism that was used to store (imperatively) the
list of affected paths. Now the patches (defined by [return], [return_payload]
and [return_paths]) store both the resulting path and the list of affected
paths.

- Added a [payload] mechanism used only for the type constraint case, where the
replacement declaration is returned as an additional payload. Other cases return
[None].

- Create separate functions ([merge_type], [merge_module], [merge_modtype])

- Extracted the post processing (wellformedness checks and substitutions) into
an helper function [post_process]

clementblaudeau committed:
[refactor merge] Extracted the recursive functions for deep constr

- Broke down the main merging function into three parts: [merge_signature],
[patch_deep_item] and [patch_all]. The first two are mutually recursive and use
an extra argument [~patch]. For now, the only patch provided (in [merge_type],
[merge_module], etc) is [patch_all].

clementblaudeau committed:
[refactor merge] Moved code

clementblaudeau committed:
[refactor merge] Extracted the type-package case

- Move the patching logic from [patch_all] to a specialized patch function in
[merge_package_constraint]

- Removed the (now) useless lid argument for [patch_all]

clementblaudeau committed:
[refactor merge] Extracted the type case

- Moved the patching logic from [patch_all] to specialized patch function in
[merge_type]

- Removed the (now) useless [initial_env] argument for [patch_all]

clementblaudeau committed:
[refactor merge] Extracted the module case

- Moved the patching logic from [patch_all] to a specialized patch function in
[merge_module]. Merged the common parts of the destructive and non-destructive
cases

clementblaudeau committed:
[refactor merge] Extracted the modtype case

- Moved the patching logic from [patch_all] to a specialized patch function in
[merge_modtype]

- Removed the (now) useless [patch_all]

clementblaudeau committed:
[refactor merge] Added comments, minor stylistic changes

- Added general comments about merging

- Renamed the `merge_package_constraint` into `merge_package`

- Moved the [try ... with] of [merge_signature] to the outer function [merge].

dra27 committed:
Add Config.c_compiler_vendor

Expose the ocaml_cc_vendor variable computed in aclocal.m4

gasche committed:
domain_create: release the domain lock on failure

damiendoligez committed:
fix #13941

damiendoligez committed:
update Changes for #13961

clef-men committed:
Add support for atomic record fields.

Requires a bootstrap.

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

clef-men committed:
Bootstrap.

To reproduce (see BOOTSTRAP.adoc for details):
- go to the earlier commit "lambda: Add support for new atomic
  primitives" as a known-good state, build that one
- keep the build artifacts around, come here and do `make bootstrap`

dra27 committed:
Dynamic matrix for the others job

Define the matrix for the others job using an actions/github-script
step, which allows the possibility of the matrix being dynamic.

dra27 committed:
Add optional static and mimimal jobs

Tests a full `--disable-shared` build on Linux and also a Linux build
with as many options disabled as possible (as the minimal build in the
other-configs job on Jenkins also does).

The matrix is expanded by adding the 'CI: Full matrix' label to a pull
request.

dra27 committed:
Merge macOS Dependency steps

dra27 committed:
Fix test_create_cursor_failures.ml on Cygwin

If Cygwin is running "elevated" - which it is in CI - then it acts as
though it's running as root. It intentionally activates
SeBackupPrivilege, which thwarts the test_create_cursor_failures.ml
test.

The OCaml testsuite will never require root privileges for anything
meaningful, so ocamltest on Cygwin simply drops the SeBackupPrivilege
when running on Cygwin, which means the test correctly fails.

dra27 committed:
Update build-msvc.yml to permit Cygwin builds

Yes, yes, the filename -msvc.yml has become a bit out of date...

clef-men committed:
Add predefined type ['a atomic_loc].

This type will be used for ['a Atomic.Loc.t], as proposed
in the RFC
  https://github.com/ocaml/RFCs/pull/39

We implement this here to be able to use it in the stdlib later,
after a bootstrap.

clef-men committed:
Add extension [%atomic.loc record.field].

clef-men committed:
Bootstrap.

This bootstrap is not required by a compiler change, but it enables
the use of the predefined type `'a atomic_loc` and the
expression-former [%atomic.loc ...] in the standard library.

gasche committed:
testsuite for pattern-matching on atomic reads

gasche committed:
allow wildcard patterns under atomic record fields

Suggested-by: Olivier Nicole <olivier@chnik.fr>

OlivierNicole committed:
Fix #13427

goldfirere committed:
Document [row_more] and [row_fixed].

gasche committed:
stdlib/Atomic: move the Loc module below in the .mli

gasche committed:
stdlib/Atomic: minimal documentation

gasche committed:
describe atomic record fields in the reference manual

MisterDA committed:
check-symbol-names: ignore extern char **environ re-decl

It is needed on Cygwin.

MisterDA committed:
gha: Document labels triggering/skipping workflows

clementblaudeau committed:
[refactor merge] Corrections from Octachron review

clementblaudeau committed:
[refactor merge] Moved the catching of exception

clementblaudeau committed:
[refactor merge] Updated comments

dra27 committed:
Windows matrix: arch instead of x86_64

dra27 committed:
Merge pull request #14013 from dra27/more-ci-tweaks

GHA: add an optional wider test matrix (Cygwin, static, minimal, etc.)

Octachron committed:
stale bot: only identify stale issues, don't close them. (#14004)

* stale bot: only identify stale issues.

bvaugon committed:
Fix memory corruption when an exception is raised during demarshaling. (#14015)

dra27 committed:
Update Changes

Octachron committed:
Merge pull request #13909 from OlivierNicole/dynarray-unsafe-to-iarray

Add Dynarray.unsafe_to_iarray

Octachron committed:
update Changes after cherry-picking

clementblaudeau committed:
[refactor merge] Added Changelog entry

Octachron committed:
Merge pull request #13911 from clementblaudeau/refactor-signature-merge

Refactor merge constraints on signatures

gasche committed:
Merge pull request #13404 from clef-men/atomic_fields_11

Atomic record fields

MisterDA committed:
Runtime events: dispatch the right event message type (#13970)

* Dispatch the right event message type

`type.runtime | type.user` does not really make sense as `.runtime`
and `.user` are two projections of the same union type.

Prevents a MSVC 19.44.35109.1 warning:

    runtime/runtime_events.c(595): warning C5287: operands are different enum types
    'ev_runtime_message_type' and 'ev_user_message_type'; use an explicit cast to silence this warning

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

* Fit in 80 cols

---------

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

nojb committed:
Add Unix.unsetenv

nojb committed:
Add test

gasche committed:
Merge pull request #13991 from gasche/atomic_fields_manual

describe atomic record fields in the reference manual

dra27 committed:
Merge pull request #14020 from nojb/unix_unsetenv

Add `Unix.unsetenv`

dra27 committed:
Merge pull request #13953 from gasche/release-domain-lock-on-spawn-failure

domain_create: release the domain lock on failure

nojb committed:
Add Option.product and Option.Syntax (#13916)

MisterDA committed:
Fix unterminated-string-initialization warning

    yacc/reader.c:62:1: error: initializer-string for character array is too long, array size is 32 but initializer has size 33 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Werror,-Wunterminated-string-initialization]
       62 | "\000\000\000\000\000\000\000\000\376\377\377\207\376\377\377\007\000\000\000\000\000\000\000\000\377\377\177\377\377\377\177\377";
          | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    yacc/reader.c:64:1: error: initializer-string for character array is too long, array size is 32 but initializer has size 33 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Werror,-Wunterminated-string-initialization]
       64 | "\000\000\000\000\200\000\377\003\376\377\377\207\376\377\377\007\000\000\000\000\000\000\000\000\377\377\177\377\377\377\177\377";
          | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

gasche committed:
[minor] More readable output from `gen_sizeclasses` (#13923)

* more readable output for tools/gen_sizeclasses

* gen_sizeclasses: document the less obvious definitions

gasche committed:
Merge pull request #14024 from MisterDA/unterminated-string-initialization

Fix unterminated-string-initialization warning

gasche committed:
[minor] use a record in Compenv.process_deferred_actions

Reviewed-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

gasche committed:
fix atomic-locs/cmm.compilers.reference

MisterDA committed:
tests: native-debugger: sanitize GDB+musl output

gasche committed:
Merge pull request #13987 from OlivierNicole/fix-tsan-false-alarm

Fix #13427

nojb committed:
Cleanups following introduction of `Pexp_struct_item`/`Texp_struct_item` (#14028)

* Remove stale comments

* Remove dead constructor: Scoping_let_module

* Remove eqparsetree.ml

goldfirere committed:
Add row_more invariant

jmid committed:
Update flexdll submodule to 0.44

jmid committed:
Reenable lib-dynlink-domains test on Windows, temporarily disabled in #11607

MisterDA committed:
tests: native-debugger: sanitize LLDB+GLIBC output

goldfirere committed:
Comments from review

goldfirere committed:
Changelog

dra27 committed:
Merge pull request #14032 from jmid/flexdll-0.44

Update FlexDLL to 0.44 and reenable lib-dynlink-domains test on Windows

damiendoligez committed:
simplifications suggested by reviewer

gasche committed:
Merge pull request #14033 from MisterDA/tests-native-debugger-gdb-musl

tests: native-debugger: sanitize GDB+musl output

gasche committed:
Merge pull request #14023 from goldfirere/document-row-more

Document [row_more] and [row_fixed].

gasche committed:
Merge pull request #13961 from damiendoligez/fix-13941-ocamltest-variables

Fix ocamltest variable handling

gasche committed:
Merge pull request #14026 from gasche/compenv-minor-refactor

[minor] use a record in Compenv.process_deferred_actions

nojb committed:
Amend docstring

dra27 committed:
Merge pull request #13728 from dra27/Sys.interpreter

Add `Sys.runtime_executable` and `caml_sys_proc_self_exe`
